### PR TITLE
Early types for the Transacter APIs

### PR DESCRIPTION
**Pull request by swankjesse**
The most interesting thing here is hooking up the Id type, which was
way more painful than it should have been.

### CODE REVIEW

File: misk/src/main/kotlin/misk/hibernate/Transacter.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package misk.hibernate
+
+interface Transacter {
+  fun <T> transaction(lambda: (session: Session) -> T): T
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/Transacter.kt]: it's so beautiful!

File: misk/src/main/kotlin/misk/hibernate/Id.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package misk.hibernate
+
+import java.io.Serializable
+
+/** Type-safe persistent identifier, mapped to a long column. */
+data class Id<T : DbEntity<T>>(val id: Long) : Serializable {
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/Id.kt]: do you think `Long` is too restrictive?

File: misk/src/main/kotlin/misk/hibernate/Session.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package misk.hibernate
+
+import org.hibernate.query.Query
+import javax.persistence.criteria.CriteriaBuilder
+import javax.persistence.criteria.CriteriaQuery
+
+interface Session {
+  fun <T : DbEntity<T>> save(entity: T)
+  fun newCriteriaBuilder(): CriteriaBuilder
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/Session.kt]: is there anything Session specific about a `Criteria`? wondering if this belongs somewhere else

File: misk/src/main/kotlin/misk/hibernate/Id.kt (hunk 2 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/Id.kt]: I think it's the right type for DB-internal row IDs. It's big enough to be unique and small enough to be efficient within MySQL.

**Comment by swankjesse** [comment]
I’ve rebased this on the DependentServices PR, which was why the tests were failing. We were racing to set up Hibernate and HSQL.

The PR is bigger! Please take a look.

File: misk/src/test/kotlin/misk/hibernate/TransacterTest.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      session.save(DbMovie("Jurassic Park", Date()))
+      session.save(DbMovie("Star Wars", Date()))
+    }
+
+    // Query some movies.
+    transacter.transaction { session ->
+      val criteriaBuilder = session.newCriteriaBuilder()
+      val criteria = criteriaBuilder.createQuery(DbMovie::class.java)
+      val queryRoot = criteria.from(DbMovie::class.java)
+      criteria.where(criteriaBuilder.notEqual(queryRoot.get<String>("name"), "Star Wars"))
```
[the remark this answers is not in the document]
Comment (swankjesse) [reply misk/src/test/kotlin/misk/hibernate/TransacterTest.kt]: Yeah, the built in one is very . . . Hibernate

File: misk/src/test/kotlin/misk/hibernate/TransacterTest.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    // Insert some movies in a transaction.
+    transacter.transaction { session ->
+      session.save(DbMovie("Jurassic Park", Date()))
+      session.save(DbMovie("Star Wars", Date()))
+    }
+
+    // Query some movies.
+    transacter.transaction { session ->
+      val criteriaBuilder = session.newCriteriaBuilder()
+      val criteria = criteriaBuilder.createQuery(DbMovie::class.java)
```
[the remark this answers is not in the document]
Comment (swankjesse) [reply misk/src/test/kotlin/misk/hibernate/TransacterTest.kt]: Gonna build a better abstraction over all of this.

File: misk/src/main/kotlin/misk/CoordinatedService.kt [hunk 6]
```diff
@@ -13,18 +14,9 @@ import com.google.inject.Key
...
-    }
-
-  val consumedKeys: Set<Key<*>>
-    get() {
-      service as? DependentService ?: return setOf()
-      return service.consumedKeys
-    }
+internal class CoordinatedService(val service: Service) : AbstractService(), DependentService {
+  override val producedKeys = (service as? DependentService)?.producedKeys ?: setOf()
+  override val consumedKeys = (service as? DependentService)?.consumedKeys ?: setOf()
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/CoordinatedService.kt]: FYI, not a behavior change!

File: misk/src/main/kotlin/misk/hibernate/HibernateModule.kt [hunk 7]
```diff
@@ -40,30 +39,35 @@ class HibernateModule(
...
 
-    val connectorKey = HibernateConnector::class.toKey(qualifier)
-    val connectorProvider = getProvider(connectorKey)
+    val sessionFactoryServiceKey = SessionFactoryService::class.toKey(qualifier)
 
     val schemaMigratorKey = SchemaMigrator::class.toKey(qualifier)
     val schemaMigratorProvider = getProvider(schemaMigratorKey)
 
-    bind(configKey).toInstance(config)
+    val transacterKey = Transacter::class.toKey(qualifier)
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/HibernateModule.kt]: so readable!

File: misk/src/main/kotlin/misk/hibernate/SchemaMigratorService.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import javax.inject.Provider
+import kotlin.reflect.KClass
+
+internal class SchemaMigratorService(
+  private val environment: Environment,
+  qualifier: KClass<out Annotation>,
+  private val schemaMigratorProvider: Provider<SchemaMigrator> // Lazy!
+) : AbstractIdleService(), DependentService {
+
+  override val consumedKeys = setOf<Key<*>>(SessionFactoryService::class.toKey(qualifier))
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/SchemaMigratorService.kt]: nice

File: misk/src/main/kotlin/misk/hibernate/SchemaMigratorService.kt [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  override val producedKeys = setOf<Key<*>>(SchemaMigratorService::class.toKey(qualifier))
+
+  override fun startUp() {
+    val schemaMigrator = schemaMigratorProvider.get()
+    if (environment == Environment.TESTING || environment == Environment.DEVELOPMENT) {
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/hibernate/SchemaMigratorService.kt]: unrelated to your PR, but I think we'll see this condition a lot. we should add something to `Environment`

File: misk/src/test/kotlin/misk/hibernate/TruncateTablesService.kt [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import javax.inject.Provider
+import kotlin.reflect.KClass
+
+/**
+ * Truncate tables before running each test.
+ *
+ * We truncate _before_ tests because that way we always have a clean slate, even if a preceding
+ * test wasn't able to clean up after itself.
+ */
+// TODO(jwilson): promote this to the misk testing library.
```
Comment (ryanhall07) [inline misk/src/test/kotlin/misk/hibernate/TruncateTablesService.kt]: do you think we're going to go down the junit rule route? or promote the idea of services for testing too?

File: misk/src/main/kotlin/misk/CoordinatedService.kt [hunk 11]
```diff
@@ -57,11 +53,11 @@ internal class CoordinatedService(val service: Service) : AbstractService() {
 
   fun startIfReady() {
     synchronized(this) {
-      if (state() != Service.State.STARTING) return
+      if (state() != State.STARTING || service.state() != State.NEW) return
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/CoordinatedService.kt]: can you update the tests to catch these state transition errors?

File: misk/src/main/kotlin/misk/hibernate/SchemaMigratorService.kt (hunk 9 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/SchemaMigratorService.kt]: Yeah, though we gotta be careful with words. `environment.isNonProduction()`!

File: misk/src/test/kotlin/misk/hibernate/TruncateTablesService.kt (hunk 10 above)
Comment (swankjesse) [reply misk/src/test/kotlin/misk/hibernate/TruncateTablesService.kt]: Services play better with injection. Also no more rules with JUnit 5!

File: misk/src/main/kotlin/misk/CoordinatedService.kt (hunk 11 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/CoordinatedService.kt]: I was thinking about how to do that. I might be able to do it without adding threads to the tests. Currently the test for this is single-threaded so it’s unlikely to be flaky.

**Pull request merged**